### PR TITLE
Make TC program application graceful

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

- [x] Never delete the qdisc.
- [x] Add new program then remove the old.
- [x] Support policy update without replacing program.

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
In BPF mode, Felix now applies policy updates without reapplying the BPF programs; this gives a performance boost and closes a window where traffic was not policed.
```

### CODE REVIEW

File: bpf/tc/attach.go [hunk 1]
```diff
@@ -104,13 +91,60 @@ func (ap AttachPoint) AttachProgram() error {
...
+		pref   string
+		handle string
+	}
+	var progsToClean []progToCleanUp
+	for _, line := range strings.Split(string(out), "\n") {
+		if !strings.Contains(line, "calico") {
+			continue
+		}
+		// find the pref and the handle
+		prefHandleRe := regexp.MustCompile(`pref ([^ ]+) .* handle ([^ ]+)`)
```
Comment (tomastigera) [inline bpf/tc/attach.go:121]: This one is a constant pattern, can be compiled just once in init time

File: bpf/tc/attach.go [hunk 2]
```diff
@@ -130,7 +164,19 @@ func (ap AttachPoint) AttachProgram() error {
...
+
+	// Success: clean up the old programs.
+	for _, p := range progsToClean {
+		log.WithField("prog", p).Debug("Cleaning up old calico program")
+		tcCmd := exec.Command("tc", "filter", "del", "dev", ap.Iface, string(ap.Hook), "pref", p.pref, "handle", p.handle, "bpf")
+		err := tcCmd.Run()
+		if err != nil {
+			// TODO: filter error to avoid spam if interface is gone.
+			log.WithError(err).WithField("prog", p).Warn("Failed to clean up old calico program.")
+			return errors.WithMessage(err, "failed to clean up old program")
```
Comment (neiljerram) [inline bpf/tc/attach.go:178]: Would it be better to try to clean up the whole of progsToClean, before returning an error?

File: fv/bpf_test.go [hunk 3]
```diff
@@ -1416,7 +1416,7 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
 								if testOpts.connTimeEnabled {
 									Skip("FIXME externalClient also does conntime balancing")
 								}
-
+								externalClient.EnsureBinary("test-connection")
```
Comment (neiljerram) [inline fv/bpf_test.go:1419]: Is this a related change?

File: dataplane/linux/bpf_ep_mgr.go [hunk 4]
```diff
@@ -496,44 +496,48 @@ func (m *bpfEndpointManager) applyPolicy(wlID proto.WorkloadEndpointID) error {
...
+			return err
 		}
-	}()
 
+		jumpMapFD, err := FindJumpMap(ap)
+		if err != nil {
+			return errors.Wrap(err, "failed to look up jump map")
+		}
+		if iface.jumpMapFD == nil {
+			iface.jumpMapFD = map[PolDirection]bpf.MapFD{}
```
Comment (neiljerram) [inline dataplane/linux/bpf_ep_mgr.go:531]: Can we simplify by doing this in the iface constructor?  (So that the map always exists.)

File: dataplane/linux/bpf_ep_mgr.go [hunk 5]
```diff
@@ -496,44 +496,48 @@ func (m *bpfEndpointManager) applyPolicy(wlID proto.WorkloadEndpointID) error {
...
+		}
+		iface.jumpMapFD[polDirection] = jumpMapFD
+		m.ifaces[endpoint.Name] = iface
+	}
+
+	return m.updatePolicyProgram(iface.jumpMapFD[polDirection], rules)
+}
+
+func (m *bpfEndpointManager) updatePolicyProgram(jumpMapFD bpf.MapFD, rules [][][]*proto.Rule) error {
```
Comment (neiljerram) [inline dataplane/linux/bpf_ep_mgr.go:540]: Is there any change needed here because of the jump map fd being held open?  E.g. if it was like a normal file, it would point to a particular offset in the file, and that might need resetting?

File: bpf/tc/attach.go (hunk 2 above)
Comment (fasaxc) [reply bpf/tc/attach.go:178]: Yes, good catch.

File: dataplane/linux/bpf_ep_mgr.go (hunk 5 above)
Comment (fasaxc) [reply dataplane/linux/bpf_ep_mgr.go:540]: No, I don't think BPF map FDs are stateful.

File: dataplane/linux/bpf_ep_mgr.go (hunk 4 above)
Comment (fasaxc) [reply dataplane/linux/bpf_ep_mgr.go:531]: Tried to do this but found a problem: we're currently using the zero value returned by `ifaces[<unknown key>]` in anger.  If I change to using a constructor, I'll have to do 
```
if iface, ok := ifaces[key]; !ok { 
   // create the entry
}
```
everywhere, or wrap that in a method.  Not sure that ends up any cleaner in the end?

File: fv/bpf_test.go (hunk 3 above)
Comment (fasaxc) [reply fv/bpf_test.go:1419]: No, looks like an accidental commit.

File: dataplane/linux/bpf_ep_mgr.go (hunk 4 above)
Comment (neiljerram) [reply dataplane/linux/bpf_ep_mgr.go:531]: OK, thanks for looking at it.
